### DISCUSSION TITLE
Angular + Source block not showing props configured in Args in story

### DISCUSSION THREAD

**Discussion by atomicrobokid** [storybookjs/storybook#28390] [Help] [answered]
### Summary

Hi,

I'm hoping you can help.

I have a story configured as follows:

```
export const Card: Story = {
  render: (args) => ({
    props: args,
    template: `<card ${argsToTemplate(args)}>
      </card>`
  }),

  args: {
    title: 'Card test title',
  }
};
```

And i've created an mdx file to display the component (which previews fine and the props update when you tinker with them)

The issue i'm having is using the `<Source>` block. It just renders the key of the prop in the preview, not the actual value itself 

Result below:
<img width="777" alt="Screenshot 2024-06-28 at 21 21 05" src="https://github.com/storybookjs/storybook/assets/72664200/eb70f826-586f-4544-b4e2-4d61d5581f3b">

This is my mdx content:
```
<Story of={CardStories.Card}></Story>

<Source of={CardStories.Card} />
```

Feel like i'm missing something obvious here, but can't figure it out!

Thanks in advance

### Additional information

Angular 17, latest SB

**Comment by atomicrobokid** [comment]
After experimenting a bit more this works as expected when you use `component: CardComponent` and don't use the `template` tag. However as a lot of my component use content projection, so i need to use the `template` tag

**Comment by atomicrobokid** [comment]
Further digging, this works:

```
template: `<card [title]="'${args.title}'"></card>`
```

That would imply `argsToTemplate` function built into storybook and/or doing `[title]="title"` doesn't behave as expected, even though the `Story` block handles it fine

**Comment by Marklb** [comment, marked as answer]
`argsToTemplate` is a simple helper function and it doesn't have access to the component data, such as what properties are inputs/outputs or the `argType` info, because it happens when that data is still being collected. Internally, Storybook is using two template translators, when a story doesn't provide it's own template. One is used for running the story and another is used for the displayed source.

You mentioned a lot of your components using content projection and I assume you are putting that projection code within the tags of your component. If your template is going to only put code that surrounds your component then I would recommend the `componentWrapperDecorator` decorator, because you can still let Storybook handle the generation of your component, while providing a custom wrapping template. I don't think that is what you need, but the docs on that is here: https://storybook.js.org/docs/writing-stories/decorators?renderer=angular

`argsToTemplate` acts like a simplified version of the props generation used for generating the template that is run in Angular. So, it should simplify creating a working story, but since you are providing your own template, Storybook is not going to touch it and try to replace anything in your template string.

There are improvements that can be made to `argsToTemplate`, for sure, if you want to contribute to it.

If you need to be able to customize both the template that is run in Angular and the template displayed as the source then that would probably be more complicated. I have an idea of what would be involved to do that, but it would be more complicated and I don't have time to work out what all would be involved for that kind of workaround at the moment. It would involve a custom decorator that regenerates the template on args change and  *maybe* a custom import of the docs addon, if the custom and original decorators conflict.

NOTE: If you do the following, it should work fine and you should not notice any difference on simple components. You will be introducing a potential negative performance impact and change to how rendering works, when changing args. When args are changed, such as making changes in the Controls addon tab, Storybook has some logic that determines if you need to fully rerender the story or just update the input property. The template being different is something that causes a full rerender. Typically Storybook is going to be running a template similar what `argsToTemplate` gives. If `args.title = 'One'` and the template is `<card [title]="title"></card>` then Controls is used to change the title to `args.title = 'Two'`, the template didn't change, so just the component input will change. If you do the following, your template would initially be `<card [title]="'One'"></card>` then after changing the Control it would be `<card [title]="'Two'"></card>`, so the component has to fully rerender for that change to happen. Most components are going to rerender fast enough that you won't even notice, but components with a long initialization or initial animations will be possibly be noticed as starting from initialization on each change.

The following is a modified `argsToTemplate`, which uses the same logic as the internal component to template inputs mapping that is used for the displayed source template.

```ts
// This is a slightly modified copy of an internal function, that isn't exported.
// Source: https://github.com/storybookjs/storybook/blob/4931b385bdcb5ba6fb9b9e2d87e0f78893fe9c42/code/frameworks/angular/src/client/angular-beta/ComputesTemplateFromComponent.ts#L67
const createAngularInputProperty = ({
  propertyName,
  value,
}: {
  propertyName: string;
  value: any;
}) => {
  let templateValue;
  switch (typeof value) {
    case 'string':
      templateValue = `'${value}'`;
      break;
    case 'object':
      // NOTE: This has a side-effect of breaking input objects that
      //   can't be stringified then parsed and still work. It also
      //   will error for objects that contain a circular reference,
      //   which is a known issue.
      templateValue = JSON.stringify(value)
        .replace(/'/g, '\u2019')
        .replace(/\\"/g, '\u201D')
        .replace(/"([^-"]+)":/g, '$1: ')
        .replace(/"/g, "'")
        .replace(/\u2019/g, "\\'")
        .replace(/\u201D/g, "\\'")
        .split(',')
        .join(', ');
      break;
    default:
      templateValue = value;
  }

  return `[${propertyName}]="${templateValue}"`;
};

// Slightly modified version of `argsToTemplate`.
// Source: https://github.com/storybookjs/storybook/blob/4931b385bdcb5ba6fb9b9e2d87e0f78893fe9c42/code/frameworks/angular/src/client/argsToTemplate.ts#L56
function argsToTemplateCustom<A extends Record<string, any>>(
  args: A,
  options: ArgsToTemplateOptions<keyof A> = {}
) {
  const includeSet = options.include ? new Set(options.include) : null;
  const excludeSet = options.exclude ? new Set(options.exclude) : null;

  return Object.entries(args)
    .filter(([key]) => args[key] !== undefined)
    .filter(([key]) => {
      if (includeSet) return includeSet.has(key);
      if (excludeSet) return !excludeSet.has(key);
      return true;
    })
    .map(([key, value]) =>
      typeof value === 'function' ? `(${key})="${key}($event)"` : createAngularInputProperty(key, args[key])
    )
    .join(' ');
}
```

**Reply by atomicrobokid** [reply]
@Marklb - thank you very much for the incredibly detailed reply, very useful and really helps me clarify what goes on under the hood of Storybook! I'll give your modified `argsToTemplate` a try but that looks like exactly what I'd be after a first glance. I'll keep in mind the performance impact when playing with some of the more complex components, and at for those i have the fallback of my 3rd post which works fine

Thanks again!
